### ISSUE TITLE
Non-deterministic results in tf_ranking_tfrecord.py

### ISSUE THREAD

**Issue by davidmosca** [tensorflow/ranking#204]
Hello,

When I run [tf_ranking_tfrecord.py](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_tfrecord.py), I get each time different nDCG metrics.

I have already tried the following:

- tf.random.set_seed(1)
- tf.compat.v1.random.set_seed(1)
- `shuffle=False` in `_input_fn()`
- and I have not modified `group_size=1` 

Is it possible to make the results deterministic? And, if so, how?

Thanks.

**Comment by xuanhuiwang** [comment]
This is a small data set. I guess the randomness may comes from the weight initialization? 

@ramakumar1729, do you have any other thoughts on the randomness?

**Comment by ramakumar1729** [comment]
@davidmosca : Which data are you using? `tensorflow_ranking/examples/data/train_elwc.tfrecord` and corresponding `test_elwc.tfrecord` are dummy datasets and should give you ndcg to be 1.0. To use the Antique dataset, I would suggest using the demo at [git.io/tf-ranking-demo](git.io/tf-ranking-demo) which downloads the Antique dataset and has more realistic metrics. The demo also shows how to generate predictions, and you should not see any non-determinism there.

**Comment by davidmosca** [comment]
Hi, I'm using the Antique dataset:
```
!wget -O "/tmp/vocab.txt" "http://ciir.cs.umass.edu/downloads/Antique/tf-ranking/vocab.txt"
!wget -O "/tmp/train.tfrecords" "http://ciir.cs.umass.edu/downloads/Antique/tf-ranking/ELWC/train.tfrecords"
!wget -O "/tmp/test.tfrecords" "http://ciir.cs.umass.edu/downloads/Antique/tf-ranking//ELWC/test.tfrecords"
```
The [Tutorial: TF-Ranking for sparse features](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/handling_sparse_features.ipynb/#) also gives non-deterministic results.

**Comment by davidmosca** [comment]
> This is a small data set. I guess the randomness may comes from the weight initialization?
> 
> @ramakumar1729, do you have any other thoughts on the randomness?

@ramakumar1729 How could I initialise weights deterministically?

**Comment by nishathussain** [comment]
@davidmosca can it be due to `approx_ndcg_loss` ??

**Comment by ramakumar1729** [comment]
@davidmosca Thanks for pointing this out, I'll have to take a closer look. In TF1, there is a notion of graph-level seed and op-level seed. `tf.set_random_seed` sets the seed at graph-level, but not at the op-level I think.

**Comment by davidmosca** [comment]
> @davidmosca can it be due to `approx_ndcg_loss` ??

@nishathussain Is `approx_ndge_loss` non-deterministic? [At first glance](https://dl.acm.org/doi/pdf/10.1145/3331184.3331347) I can't see anything to suggest so.

**Comment by vkaul11** [comment]
This is an issue we are also facing especially when we ported our tf 1.0 code to tf 2.0. I had a question do you intend to port tensorflow ranking code to native tf 2.0 without using "tf.compat.v1."? We were using tf lattice models and they recommend using v2 optimizers like they do in this code https://github.com/tensorflow/lattice/blob/master/examples/custom_estimators_uci_heart.py optimizers.get_optimizer_instance_v2('Adam',
                                                       FLAGS.learning_rate).

**Comment by davidmosca** [comment]
@vkaul11 Ideally I'd prefer to use native TF 2.0 code (since it's the current version) although I'm not sure how straightforward it is to port the TFR code to TF 2.0. Either way, as long as the code can be run deterministically, I'm fine. Thanks.

**Comment by vkaul11** [comment]
@ramakumar1729 is there a plan to move TFR code to TF 2.0 native and use the new optimizers?

**Comment by ramakumar1729** [comment]
In release [v0.3.0](https://github.com/tensorflow/ranking/releases/tag/v0.3.0), we released support for Keras/TF2 APIs. We do not have a native keras trainer demo yet (though we do support it), but use model_to_estimator to be compatible with TF1 Estimator workflows. Note that the keras model in the [tf_ranking_keras_tfrecord.py](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/keras/keras_m2e_tfrecord.py#L204) used TF2 optimizers.

**Comment by vkaul11** [comment]
@ramakumar1729 I was trying to port tf 1.0 to tf 2.0 code 

`def train_and_eval():

  """Train and Validate."""



  features, labels = load_libsvm_data(FLAGS.dataset_base_path + '/train-' + FLAGS.locale + '.txt', 10)

  train_input_fn, train_hook = get_train_inputs(features, labels, FLAGS.batch_size)



  features_vali, labels_vali = load_libsvm_data( FLAGS.dataset_base_path + '/test-' + FLAGS.locale + '.txt', 10)

  vali_input_fn, vali_hook = get_eval_inputs(features_vali, labels_vali)



  optimizer = tf.compat.v1.train.AdagradOptimizer(learning_rate=FLAGS.learning_rate)



  def _train_op_fn(loss):

    """Defines train op used in ranking head."""

    update_ops = tf.compat.v1.get_collection(tf.compat.v1.GraphKeys.UPDATE_OPS)

    minimize_op = optimizer.minimize(

        loss=loss, global_step=tf.compat.v1.train.get_global_step())

    train_op = tf.group([minimize_op, update_ops])

    return train_op



  ranking_head = tfr.head.create_ranking_head(

        loss_fn=tfr.losses.make_loss_fn(tfr.losses.RankingLossKey.APPROX_NDCG_LOSS),

        eval_metric_fns=eval_metric_fns(),

        train_op_fn=_train_op_fn)



  estimator = tf.estimator.Estimator(

    model_fn=tfr.model.make_groupwise_ranking_fn(

      group_score_fn=make_score_fn(),

      group_size=FLAGS.group_size,

      transform_fn=make_transform_fn(),

      ranking_head=ranking_head),

      config=tf.estimator.RunConfig(

          FLAGS.output_dir, save_checkpoints_steps=1000))



  train_spec = tf.estimator.TrainSpec(

    input_fn=train_input_fn,

    hooks=[train_hook],

    max_steps=FLAGS.num_train_steps)

  vali_spec = tf.estimator.EvalSpec(

    name="eval",

    input_fn=vali_input_fn,

    hooks=[vali_hook],

    #steps=10000,

    throttle_secs=15)`

Here **group_score_fn=make_score_fn()** we use the https://github.com/tensorflow/lattice model score output. But Lattice guys recommended we use v2 optimizers. Do you recommend that instead of using  tf.compat.v1.train.AdagradOptimizer and tf.compat.v1.train.AdagradOptimizer we use model_to_estimator pipeline that you wrote about in order to use v2 optimizers?

**Comment by nishathussain** [comment]
@davidmosca one more thing I can think of is `tf_random_seed` value in runConfig.

tf.estimator.RunConfig(
    model_dir=None, tf_random_seed=None, save_summary_steps=100,
    save_checkpoints_steps=_USE_DEFAULT, save_checkpoints_secs=_USE_DEFAULT,
    session_config=None, keep_checkpoint_max=5, keep_checkpoint_every_n_hours=10000,
    log_step_count_steps=100, train_distribute=None, device_fn=None, protocol=None,
    eval_distribute=None, experimental_distribute=None,
    experimental_max_worker_delay_secs=None, session_creation_timeout_secs=7200
)

https://www.tensorflow.org/api_docs/python/tf/estimator/RunConfig

**Comment by davidmosca** [comment]
@nishathussain Adding the `tf_random_seed`to RunConfig did the job - the results are now deterministic. Many thanks.
